src/app/OperationalSessionSetup.h

```diff
@@ -155,6 +155,34 @@ typedef void (*OnDeviceConnectionRetry)(void * context, const ScopedNodeId & pee
 class DLL_EXPORT OperationalSessionSetup : public SessionEstablishmentDelegate, public AddressResolve::NodeListener
 {
 public:
+    enum class State : uint8_t
+    {
+        Uninitialized,    // Error state: OperationalSessionSetup is useless
+        NeedsAddress,     // No address known, lookup not started yet.
+        ResolvingAddress, // Address lookup in progress.
+        HasAddress,       // Have an address, CASE handshake not started yet.
+        Connecting,       // CASE handshake in progress.
+        SecureConnected,  // CASE session established.
+        WaitingForRetry,  // No address known, but a retry is pending.  Added at
+                          // end to make logs easier to understand.
+    };
+
+    struct ConnnectionFailureInfo
+    {
+        const ScopedNodeId & peerId;
+        CHIP_ERROR error;
+        OperationalSessionSetup::State sessionState;
```

Comment:
OK, so this comes back to the question I asked in the other PR: does exposing this _internal state_ give consumers what they actually want?  I am not sure it does.

Reply:
The state provides valuable context about the error, pinpointing the precise conditions under which the failure occurs.

This information is crucial for our statistical analysis, as it allows us to track the occurrence of errors within our test matrix. Therefore, it's essential that we log the state simultaneously when recording the error.

I have transfered all the comprehensive modifications for OperationalSessionSetup here to provide a full context, this PR encompasses the fundamental alterations for this Feature Request (FR). Once we reach a consensus on OperationalSessionSetup, the other PS request should undergo a swift review.

Reply:
So I think we actually want more state, i.e. we want the current CASE/PASE state

@bzbarsky-apple to address your concern: as you undoubtedly know, making the first connection to the device is where everything can go wrong in practice - router can be misconfigured, router can have bugs (e.g. TP-Link has IPv6 routing bugs), mDNS can have misconfiguration/bugs (e.g. advertising wrong address)

Effectively, a large % of the time, we're going to fail as we try to connect to the device - once we start talking to it, there's a high chance of success

As such, the more information we can get at this stage, the better, to try to help/guide the user as to what may be wrong & how they can fix it, as well as to improve our code to properly handle these cases

Reply:
@gmarcosb I am not sure how this addresses my concern....

Again, my concern is that in the current form this setup cannot tell apart "We never resolved an IP address at all" and "We resolved an IP address, but Sigma1 sent to that address never got a response".

Reply:
Instead of expose any internal state, Introduced a new Session State to augment errorCode instead.